tests/phpunit/CRM/ManualDirectDebit/Hook/CalculateContributionReceiveDate/FirstContributionTest.php

```diff
@@ -125,4 +125,70 @@ public function testPaymentPlansNotPaidWithDirectDebitAreNotChanged() {
     $this->assertEquals('2020-01-15 00:00:00', $receiveDate);
   }
 
+  public function testReceiveDateCalculationScenarios() {
+    $scenarios = [
+      'IO131TestScenario' => ['2020-12-18 00:00:00', 3, [4], [5, 21], '2021-01-21 00:00:00'],
+      'StartDateBeforeNIRD' => ['2020-08-02 00:00:00', 3, [4], [5, 21], '2020-08-21 00:00:00'],
+      'SignUpAfterNIRD' => ['2020-08-04 00:00:00', 3, [4], [5, 21], '2020-09-21 00:00:00'],
+      'OnPaymentRunDate' => ['2020-08-04 00:00:00', 3, [4], [5, 21], '2020-09-21 00:00:00'],
+      'Scenario5' => ['2020-09-09 00:00:00', 10, [5, 10], [5, 25], '2020-09-25 00:00:00'],
+      'Scenario6' => ['2020-09-11 00:00:00', 15, [5], [25], '2020-10-25 00:00:00'],
+      'Sheela#1' => ['2020-09-10 00:00:00', 10, [5, 10], [5, 25], '2020-10-25 00:00:00'],
+      'Sheela#2' => ['2020-08-01 00:00:00', 3, [4], [5, 21], '2020-08-21 00:00:00'],
+      'Sheela#3' => ['2020-08-01 00:00:00', 3, [4], [7, 21], '2020-08-21 00:00:00'],
+      'Sheela#4' => ['2020-08-05 00:00:00', 3, [4, 18], [7, 21], '2020-09-07 00:00:00'],
+      'Sheela#5' => ['2020-08-20 00:00:00', 3, [4, 18], [5, 21], '2020-09-21 00:00:00'],
+      'Sheela#6' => ['2020-02-04 00:00:00', 3, [4], [5, 21], '2020-03-21 00:00:00'],
+    ];
+
+    foreach ($scenarios as $scenarioName => $testData) {
+      $startDate = $testData[0];
+      $minDaysToFirstPayment = $testData[1];
+      $newInstructionDates = $testData[2];
+      $paymentCollectionRunDates = $testData[3];
+      $expectedReceiveDateForFirstContribution = $testData[4];
+      $calculatedReceiveDate = $this->calculateReceiveDate($startDate, $minDaysToFirstPayment, $newInstructionDates, $paymentCollectionRunDates);
+
+      $this->assertEquals(
+        $expectedReceiveDateForFirstContribution,
+        $calculatedReceiveDate,
+        "Scenario $scenarioName failed!"
+      );
+    }
+  }
+
+  /**
+   * Calculates the receive date for the first contribution.
+   *
+   * @param string $startDate
+   * @param int $minDaysToFirstPayment
+   * @param array $newInstructionDates
+   * @param array $paymentCollectionRunDates
+   *
+   * @return mixed
+   * @throws \CRM_Extension_Exception
+   * @throws \CiviCRM_API3_Exception
+   */
+  private function calculateReceiveDate($startDate, $minDaysToFirstPayment, $newInstructionDates, $paymentCollectionRunDates) {
```

Comment:
@MiyaNoctem  should this a separate function for calculate receive date. I believe the logic for calculate received date should be in testReceiveDateCalculationScenarios test as it has not used anywhere else. However, if you wish to decouple, perhaps, the mocking setting manager should be a separate function as it is repeated in every test.